microsoft/terminal
# Enable setting an initial position and maximization launch for Terminal

This PR is for code changes as well as a spec for https://github.com/microsoft/terminal/issues/1043

 
No


## PR Checklist
* [x] Closes #1043 


This PR includes the code changes that enable users to set an initial position (top left corner) and launch maximized. There are some corner cases:

1. Multiple monitors. The user should be able to set the initial position to any monitors attached. For the monitors on the left side of the major monitor, the initial position values are negative. 

2. If the initial position is larger than the screen resolution and the window is off-screen, the current solution is to check if the top left corner of the window intersect with any monitors. If it is not, we set the initial position to the top left corner of the nearest monitor.

3.  If the user wants to launch maximized and provides an initial position, we launch the maximized window on the monitor where the position is located. 


Testing:

To test:
1. Check-out this branch and build on VS2019
2. Launch Terminal, and open Settings. Then close the terminal. 
3. Add the following setting into Json settings file as part of "globals", just after "initialRows":
                   "initialPosition": "1000, 1000",
                   "launchMode": "default"

My test data: 
I have already tested with the following variables:
                                1. showTabsInTitlebar true or false
                                2. The initial position of the top left corner of the window
                                3. Whether to launch maximized
                                4. The DPI of the monitor

Test data combination:

Non-client island window (showTabsInTitlebar true)

1. Three monitors with the same DPI (100%), left, middle and right, with the middle one as the primary, resolution: 1980 * 1200, 1920 * 1200, 1920 * 1080
    launchMode: default
                          In-Screen test: (0, 0), (1000, 500), (2000, 300), (-1000, 400), (-100, 200), (-2000, 100),
                                                  (0, 1119)
                          out-of-screen:
                                                  (200, -200): initialize to (0, 0)
                                                  (200, 1500): initialize to (0, 0)
                                                  (2000, -200): initialize to (1920, 0)
                                                  (2500, 2000): initialize to (1920, 0)
                                                  (4000 100): initialize to (1920, 0)
                                                  (-1000, -100): initialize to (-1920, 0)
                                                  (-3000, 100): initialize to (-1920, 0)
                                                  (10000, -10000): initialize to (1920, 0)
                                                  (-10000, 10000): initialize to (-1920, 0)
                                                  (0, -10000): initialize to (0, 0)
                                                  (0, -1):  initialize to (0, 0)
                                                  (0, 1200):  initialize to (0, 0)
    launch mode: maximize
                                   (100, 100)
                                   (-1000, 100): On the left monitor
                                   (0, -2000): On the primary monitor
                                   (10000, 10000): On the primary monitor


2. Left monitor 200% DPI, primary monitor 100% DPI
    In screen: (-1900, 100), (-3000, 100), (-1000, 100)
    our-of-screen: (-8000, 100): initialize at (-1920, 0)
    launch Maximized:  (-100, 100): launch maximized on the left monitor correctly

3. Left monitor 100% DPI, primary monitor 200% DPI
    In-screen: (-1900, 100), (300, 100), (-800, 100), (-200, 100)
    out-of-screen: (-3000, 100): initialize at (-1920, 0)
    launch maximized: (100, 100), (-1000, 100)

For client island window, the test data is the same as above. 

Issues:

1. If we set the initial position on the monitor with a different DPI as the primary monitor, and the window "lays" across two monitors, then the window still renders as it is on the primary monitor. The size of the window is correct.
